enhancements/network/ingress-node/ingress-node-firewall-with-ebpf.md

```diff
@@ -0,0 +1,355 @@
+
+### Implementation Details/Notes/Constraints
+
+- sample `GlobalNodeFirewall` object
+```yaml
+kind: GlobalNodeFirewall
+metadata:
+  name: worker-nodes-firewall
+Spec:
+  selector: has(node-role.kubernetes.io/worker)
+  interface:
+  - “eth0”
+  - “eth1”
+  ingress:
+  - fromCIDRS:
+       - 1.1.1.1/24
+       - 100:1::1/64
+    protocol:
+    - tcp:
+       port: 100
+       action: Allow
+    - udp:
+       port: 80
+       action: Deny
+  - fromCIDRS:
+       - 3.3.3.3/24
+    protocol:
+    - icmp:
+      type: 3 # Destination unreachable
+      code: 1 # Host unreachable
+      action: Allow
+  - fromCIDRS:
+       - 10:10::1/64
+    protocol:
+    icmpv6:
+      type: 128 # Ping request
+      action: Deny
+```
+
+- eBPF data structures and types
+```c
+#define DENY XDP_DROP
+#define ALLOW XDP_PASS
+
+struct ruleInfo {
+    __u32 ruleId;
+    __u8 protocol;
+    __u32 srcAddr;
+    __u32 srcMask;
+    __u16 dstPort;
+    __u8 icmpType;
+    __u8 icmpCode;
+    __u8 action;
+} __packed;
+
+#define MAX_INSIDE_TARGETS   (1024)
+#define MAX_RULES_PER_TARGET (100)
+
+struct bpf_map_def SEC("maps") ingressRulesTbl = {
+    .type = BPF_MAP_TYPE_LPM_TRIE,
+    .key_size = sizeof(struct bpf_lpm_ip_trie),
+    .value_size = sizeof(struct rulesStore) + (sizeof(struct ruleInfo)*MAX_RULES_PER_TARGET),
+    .max_entries = MAX_INSIDE_TARGETS,
+    .map_flags = BPF_F_NO_PREALLOC,
+};
+
+// using Longest perfix match in case of overlapping CIDRs we need to match to the more specific CIDR.
+struct bpf_lpm_trie_ip_key {
+    __u32 prefixlen;
+    union {
+        __u8 data[4];
+        __u8 data[16];
+    } u;
+} __packed;
+
+struct rulesStore {
+    __u32 numRules;
+    struct ruleInfo rulesInfo[0];
+} __packed;
+```
+
+![eBPF datastructures](./ebpf-structures.png)
+
+### User Stories
+
+#### Story 1
+
+### API Extensions
+
+A new cluster scoped CRD is introduced to support this feature this CRD will be managed by `ClusterNetworkOperator`
+```golang
+type IngressNodeFirewallFromCIDR struct {
+	// FromCIDR is the IPv4/IPv6 CIDR from which we apply node firewall rule
+	// +kubebuilder:validation:Pattern=`^(([0-9]|[0-9][0-9]|1[0-9][0-9]|2[0-4][0-9]|25[0-5])\.){3}([0-9]|[0-9][0-9]|1[0-9][0-9]|2[0-4][0-9]|25[0-5])/([0-9]|[12][0-9]|3[0-2])$`
+	FromCIDR string `json:"fromCIDR" protobuf:"bytes,1,opt,name=fromCIDR"`
+}
+
+// IngressNodeFirewallProtocolRule define ingress node firewall rule per protocol
+type IngressNodeFirewallProtocolRule struct {
+	// Port is the port number to which clients should send traffic to.
+	// +kubebuilder:validation:Required
+	// +kubebuilder:validation:Maximum:=65535
+	// +kubebuilder:validation:Minimum:=1
+	Port int32 `json:"port" protobuf:"varint,1,opt,name=port"`
+
+	// IcmpType define ICMP Type Numbers (RFC 792).
+	// +kubebuilder:validation:Required
+	// +kubebuilder:validation:Maximum:=43
+	// +kubebuilder:validation:Minimum:=0
+	IcmpType int8 `json:"icmpType" protobuf:"varint,2,opt,name=icmpType"`
```

Comment:
should not have a single "rule" type with fields that only apply to specific types. There should be one struct with fields for ICMP matches and one for fields with UDP/TCP/SCTP matches.

Reply:
something like this ?
```golang
// IngressNodeFirewallICMPRule define ingress node firewall rule for ICMP and ICMPv6 protocols
type IngressNodeFirewallICMPRule struct {
	// ICMPType define ICMP Type Numbers (RFC 792).
	// +kubebuilder:validation:Required
	// +kubebuilder:validation:Maximum:=43
	// +kubebuilder:validation:Minimum:=0
	ICMPType int8 `json:"icmpType" protobuf:"varint,1,opt,name=icmpType"`

	// ICMPCode define ICMP Code ID (RFC 792).
	// +kubebuilder:validation:Required
	// +kubebuilder:validation:Maximum:=16
	// +kubebuilder:validation:Minimum:=0
	ICMPCode int8 `json:"icmpCode" protobuf:"varint,2,opt,name=icmpCode"`
}

// IngressNodeFirewallProtoRule define ingress node firewall rule for TCP, UDP and SCTP protocols
type IngressNodeFirewallProtoRule struct {
	/ Port is the port number to which clients should send traffic to.
	// +kubebuilder:validation:Required
	// +kubebuilder:validation:Maximum:=65535
	// +kubebuilder:validation:Minimum:=1
	Port int32 `json:"port" protobuf:"varint,1,opt,name=port"`
}

// IngressNodeFirewallProtocolRule define ingress node firewall rule per protocol
type IngressNodeFirewallProtocolRule struct {
    // IngressNodeFirewallProtoRule define ingress node firewall rule for TCP, UDP and SCTP protocols 
	ProtocolRule IngressNodeFirewallProtoRule json:"protoRule" protobuf:"varint,1,opt,name=protoRule"`
    
	// IngressNodeFirewallICMPRule define ingress node firewall rule for ICMP and ICMPv6 protocols
    ICMPRule IngressNodeFirewallICMPRule `json:"icmpRule" protobuf:"varint,2,opt,name=icmpRule"`

	// Protocol can be ICMP, ICMPv6, TCP, SCTP or UDP.
	// +kubebuilder:validation:Required
	// +kubebuilder:validation:Enum="ICMP";"ICMPv6";"TCP";"UDP";"SCTP"
	Protocol IngressNodeFirewallRuleProtocolType `json:"protocol" protobuf:"bytes,3,opt,name=protocol,casttype=IngressNodeFirewallRuleProtocolType"`

	// Action can be Allow or Deny.
	// +kubebuilder:validation:Required
	// +kubebuilder:validation:Enum="Allow";"Deny"
	Action IngressNodeFirewallActionType `json:"action" protobuf:"bytes,3,opt,name=action,casttype=IngressNodeFirewallActionType"`
}
```

Reply:
yeah... although that's ugly :slightly_smiling_face:. Need to think about this...